### PR TITLE
Implement Pointers2 package

### PR DESCRIPTION
**Pull request by kraiouchkine**
## Description

Implements the C Coding Standards Pointers2 package

closes github/coding-standards-team#763

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [x] Queries have been added for the following rules:
     - ARR39-C
 - [ ] Queries have been modified for the following rules:
     - _rule number here_

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [ ] Yes
  - [x] No

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [ ] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [ ] Have all the relevant rule package description files been checked in?
 - [ ] Have you verified that the metadata properties of each new query is set appropriately?
 - [ ] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [ ] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [ ] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [ ] Does the query have an appropriate level of in-query comments/documentation?
 - [ ] Have you considered/identified possible edge cases?
 - [ ] Does the query not reinvent features in the standard library?
 - [ ] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: c/cert/test/rules/ARR39-C/test.c [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+struct s1 {
+  int v1, v2, v3;
+};
+
+void f2(int p1, int p2) {
+  int *v1;
+  v1 += p1; // NON_COMPLIANT
+  v1 += p2; // COMPLIANT
+}
+
```
Comment (jsinglet) [inline c/cert/test/rules/ARR39-C/test.c:12]: In the rule I think the gist of it is that pointer arithmetic should only be allowed if the arithmetic concerns the thing the pointer points to. In the above example `*v1` doesn't point to anything, so there should be no valid pointer arithmetic allowed if I am understanding correctly.

**Review by jsinglet — CHANGES_REQUESTED** [review]
Looking good @kraiouchkine! I have a question about the test case -- can you take a look and let me know?

File: c/cert/test/rules/ARR39-C/test.c (hunk 1 above)
Comment (kraiouchkine) [reply c/cert/test/rules/ARR39-C/test.c:12]: ARR39-C does not differentiate between results based on the underlying value of the pointer. The rule only generally prohibits performing pointer arithmetic with a scaled integer operand. You are right; the test case contains undefined behavior, such as that pointer arithmetic on an uninitialized variable, but none of it falls under the scope of this rule or affects the tests.

Comment (jsinglet) [reply c/cert/test/rules/ARR39-C/test.c:12]: Makes sense, thank you! I had a closer look and I agree.

**Review by jsinglet — APPROVED** [review]
Looks good @kraiouchkine! Thanks for this!
